File: python_pure/tests/test_random_source.py
from __future__ import annotations

import pytest
from volarb_py.random_source import (
    InvalidRandomSourceError,
    advanced,
    next_bits,
    next_uniform,
    rotate_right,
    seeded_source,
    standard_normals,
    uniforms,
)

REFERENCE_STREAM = [
    9705778491962043240,
    1370407407632858425,
    11774395822783136600,
    17944889938176486912,
    14437308781460811564,
    6944869453235589526,
]
SAMPLE_SIZE = 200_000
WORD_BITS = 64
SHORT_DRAW = 3
LONGER_DRAW = 4
LOW_TAIL = 0.001
HIGH_TAIL = 0.999


def test_the_seeded_stream_matches_the_reference_implementation() -> None:
    current = seeded_source(42, 54)
    for expected in REFERENCE_STREAM:
        current, bits = next_bits(current)
        assert bits == expected


def test_rotation_wraps_rather_than_discarding_bits() -> None:
    assert rotate_right(1, 0) == 1
    assert rotate_right(1, 1) == 1 << (WORD_BITS - 1)
    assert rotate_right((1 << WORD_BITS) - 1, 37) == (1 << WORD_BITS) - 1


def test_uniforms_land_in_the_half_open_unit_interval() -> None:
    _, values = uniforms(seeded_source(1, 1), SAMPLE_SIZE)
    assert all(0.0 <= value < 1.0 for value in values)
    assert sum(values) / len(values) == pytest.approx(0.5, abs=0.005)
    assert min(values) < LOW_TAIL
    assert max(values) > HIGH_TAIL


def test_standard_normals_have_the_moments_they_claim() -> None:
    _, values = standard_normals(seeded_source(7, 3), SAMPLE_SIZE)
    mean = sum(values) / len(values)
    variance = sum((value - mean) ** 2 for value in values) / (len(values) - 1)
    assert mean == pytest.approx(0.0, abs=0.01)
    assert variance == pytest.approx(1.0, abs=0.02)


def test_different_sequences_are_different_streams() -> None:
    _, one = uniforms(seeded_source(1, 1), 64)
    _, other = uniforms(seeded_source(1, 2), 64)
    assert one != other

    _, repeated = uniforms(seeded_source(1, 1), 64)
    assert one == repeated


def test_the_state_advances_deterministically() -> None:
    source = seeded_source(11, 13)
    assert advanced(source).state == advanced(source).state
    assert advanced(source).increment == source.increment
    assert next_uniform(source)[1] == next_uniform(source)[1]


def test_a_negative_count_is_rejected() -> None:
    with pytest.raises(InvalidRandomSourceError):
        uniforms(seeded_source(1, 1), -1)
    with pytest.raises(InvalidRandomSourceError):
        standard_normals(seeded_source(1, 1), -1)
    with pytest.raises(InvalidRandomSourceError):
        seeded_source(-1, 1)
    assert uniforms(seeded_source(1, 1), 0)[1] == []


def test_an_odd_normal_count_still_consumes_whole_pairs() -> None:
    _, three = standard_normals(seeded_source(5, 5), SHORT_DRAW)
    _, four = standard_normals(seeded_source(5, 5), LONGER_DRAW)
    assert len(three) == SHORT_DRAW
    assert len(four) == LONGER_DRAW
    assert three == four[:SHORT_DRAW]

File: cpp_core/tests/test_random_source.cpp
#include "volarb/random_source.hpp"

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>

#include <cmath>
#include <set>
#include <vector>

using Catch::Approx;
using volarb::advanced;
using volarb::InvalidRandomSourceError;
using volarb::next_bits;
using volarb::next_uniform;
using volarb::PcgState;
using volarb::rotate_right;
using volarb::seeded_source;
using volarb::standard_normals;
using volarb::uniforms;

namespace {

const int sample_size = 200000;

}

TEST_CASE("the seeded stream matches the reference implementation", "[random_source]") {
    const PcgState source = seeded_source(42, 54);
    const std::vector<std::uint64_t> expected = {
        9705778491962043240ULL, 1370407407632858425ULL, 11774395822783136600ULL,
        17944889938176486912ULL, 14437308781460811564ULL, 6944869453235589526ULL,
    };
    PcgState current = source;
    for (const std::uint64_t want : expected) {
        const auto drawn = next_bits(current);
        current = drawn.state;
        REQUIRE(drawn.bits == want);
    }
}

TEST_CASE("rotation wraps rather than discarding bits", "[random_source]") {
    REQUIRE(rotate_right(1ULL, 0) == 1ULL);
    REQUIRE(rotate_right(1ULL, 1) == (1ULL << 63));
    REQUIRE(rotate_right(0xFFFFFFFFFFFFFFFFULL, 37) == 0xFFFFFFFFFFFFFFFFULL);
}

TEST_CASE("uniforms land in the half open unit interval", "[random_source]") {
    const std::vector<double> values = uniforms(seeded_source(1, 1), sample_size);
    double total = 0.0;
    double lowest = 1.0;
    double highest = 0.0;
    for (const double value : values) {
        REQUIRE(value >= 0.0);
        REQUIRE(value < 1.0);
        total += value;
        lowest = std::min(lowest, value);
        highest = std::max(highest, value);
    }
    REQUIRE(total / static_cast<double>(values.size()) == Approx(0.5).margin(0.005));
    REQUIRE(lowest < 0.001);
    REQUIRE(highest > 0.999);
}

TEST_CASE("standard normals have the moments they claim", "[random_source]") {
    const std::vector<double> values = standard_normals(seeded_source(7, 3), sample_size);
    double total = 0.0;
    for (const double value : values) {
        total += value;
    }
    const double mean = total / static_cast<double>(values.size());
    double squared = 0.0;
    for (const double value : values) {
        squared += (value - mean) * (value - mean);
    }
    const double variance = squared / static_cast<double>(values.size() - 1);
    REQUIRE(mean == Approx(0.0).margin(0.01));
    REQUIRE(variance == Approx(1.0).margin(0.02));
}

TEST_CASE("different sequences are different streams", "[random_source]") {
    const std::vector<double> one = uniforms(seeded_source(1, 1), 64);
    const std::vector<double> other = uniforms(seeded_source(1, 2), 64);
    REQUIRE(one != other);

    const std::vector<double> repeated = uniforms(seeded_source(1, 1), 64);
    REQUIRE(one == repeated);
}

TEST_CASE("the state advances deterministically", "[random_source]") {
    const PcgState source = seeded_source(11, 13);
    const PcgState once = advanced(source);
    REQUIRE(advanced(source).state == once.state);
    REQUIRE(once.increment == source.increment);
    REQUIRE(next_uniform(source).value == next_uniform(source).value);
}

TEST_CASE("a negative count is rejected", "[random_source]") {
    REQUIRE_THROWS_AS(uniforms(seeded_source(1, 1), -1), InvalidRandomSourceError);
    REQUIRE_THROWS_AS(standard_normals(seeded_source(1, 1), -1), InvalidRandomSourceError);
    REQUIRE(uniforms(seeded_source(1, 1), 0).empty());
}

TEST_CASE("an odd normal count still consumes whole pairs", "[random_source]") {
    const std::vector<double> three = standard_normals(seeded_source(5, 5), 3);
    const std::vector<double> four = standard_normals(seeded_source(5, 5), 4);
    REQUIRE(three.size() == 3);
    REQUIRE(four.size() == 4);
    for (std::size_t index = 0; index < three.size(); ++index) {
        REQUIRE(three[index] == four[index]);
    }
}
